lib/ansible/galaxy/api.py

```diff
@@ -450,13 +450,20 @@ def wait_import_task(self, task_url, timeout=0):
         state = 'waiting'
         data = None
 
-        display.display("Waiting until Galaxy import task %s has completed" % task_url)
+        # v3 returns relative task urls, v2 returns full urls
+        if 'v3' in self.available_api_versions:
```

Comment:
A few things that I have come to while looking over this:

1. I'm not a fan of special casing things like this, it fragile and will ultimately lead to the requirement of making it more complex over time
1. We should do this in a non-special cased way, such as a helper function that ensures that a URL is always fully qualified. Creating a helper of this generic functionality in `ansible.module_utils.urls` would be a great idea.  It could accept `potentially_not_fq_url` and a response object, and handle the 3 scenarios of an already FQ URL, an absolute URL path, and a relative URL path.
1. I don't believe this work should be done in `wait_import_task`, it's pushing responsibility of URL munging into a method with a well defined goal.  `wait_import_task` should be given usable data.
1. I think this work should happen as close to where the URL is obtained from, the farther away we do this from that point, the farther from the original context we are.  How do we know that the resultant URL that generated the request is the same as `self.api_server`?  Maybe users have a need for something like this, and we should be using `response.get_url()` as obtained from urllib, to determine how to best build the fully qualified URL.
1. If we have the potential of only getting back non FQ urls from AH, are there other places where this is a concern?  Such as an artifact URL?

Reply:
Also to tack this on, I really feel the API should be responsible for this, as opposed to pushing this off to clients responsibility. Potentially the assumption is being made that only `ansible-galaxy` will ever have to suffer this burden, but it's an API, and others who may interact with it will the be burdened. Consolidating this functionality into the API is less obtrusive, and less burdensome on clients of the API.

Reply:
@sivel  I took a look at what it would look like to do this in the caller last night.  In the end I discarded that approach as it seemed to break the concept of GalaxyAPI as an abstraction layer for galaxy versions.  We end up needing to push knowledge of what urls look like in v2 and v3 outside of the GalaxyAPI class and into the caller.

(I'll post what that would have looked like in a moment)

Reply:
``` diff
diff --git a/lib/ansible/galaxy/collection.py b/lib/ansible/galaxy/collection.py
index caa22308eb..52272cde52 100644
--- a/lib/ansible/galaxy/collection.py
+++ b/lib/ansible/galaxy/collection.py
@@ -381,6 +381,12 @@ def publish_collection(collection_path, api, wait, timeout):
     :param timeout: The time in seconds to wait for the import process to finish, 0 is indefinite.
     """
     import_uri = api.publish_collection(collection_path)
+    if 'v3' in api.available_api_versions:
+        parsed = urlparse(api.api_server)
+        import_uri = parsed._replace(path=import_uri).geturl()
+    else:
+        import_uri = import_uri
+
     if wait:
         display.display("Collection has been published to the Galaxy server %s %s" % (api.name, api.api_server))
         with _display_progress():
```

Reply:
That diff is a light reworking of this PR.  When I made it I saw that it still had most of the problems of this one (to be fair, we probably can merge my changes with that initial attempt and fix some of those problems) and additionally split the responsibilty for compatibility between GalaxyAPI and the caller.  So I discarded that and pushed most of the changes inside of GalaxyAPI instead.  I did parse the data from the API and normalize it (as a task_id) before sending it into the class, though.  That seemed like the best way to split the responsibility....  The caller ends up knowing about the information that was returned to it.  The GalaxyAPI class knows how to turn the normalized data into a url on each of v2 and v3.